### PR TITLE
feat(payments): add complete Razorpay orchestration and provider support

### LINKED ISSUE #1368: [Feature]: Add Razorpay as a payment provider
### What feature do you want?

Currently InsForge only supports Stripe as a payment provider in the Payments 
section. I'd like to add Razorpay as a second provider option.

The expected behavior would mirror Stripe:
- Razorpay API key configuration (Test + Live)
- Sync products/prices/customers from Razorpay
- Webhook configuration for payment events

I have already submitted PR #1347, #1382  with the initial database migration 
(razorpay_connections table), but I understand there is orchestration work 
needed beyond just the schema. I'd love guidance on the right architecture 
before proceeding further.

### Why do you need this?

Razorpay is the dominant payment gateway in India and South/Southeast Asia, 
where Stripe has limited availability or requires additional setup. Many 
developers building projects on InsForge are based in these regions and 
currently cannot use the Payments feature at all because Stripe is not 
accessible to them.

Adding Razorpay support would make InsForge's Payments module usable for a 
large portion of the developer community that is currently excluded.

### PR DESCRIPTION
**Pull request by hemanthreddykoduru**
## Summary

This PR fully implements Razorpay as a primary payment provider alongside Stripe. It builds upon the initial database schema to provide a complete backend orchestration layer, syncing mechanism, and frontend settings UI, addressing previous feedback about the need for full orchestration.

**Key Additions:**
* **Core Services:** Implemented `RazorpayProvider` and `RazorpaySyncService` for full orchestration.
* **Data Syncing:** Successfully pulls down the Razorpay Catalog (plans/items), Customers, Subscriptions, and Payment History using the exact same robust patterns established by the Stripe integration.
* **Security & Webhooks:** Added secure storage for Live and Test environment keys via `SecretService`, and implemented `RazorpayWebhookService` for real-time event synchronization.
* **Frontend UI:** Integrated Razorpay into the Settings dashboard to easily configure keys and trigger manual syncs.
* **Data Tables:** Updated the frontend tables for Payment History, Catalog, Customers, and Subscriptions to gracefully separate and display Razorpay data alongside Stripe data with matching UI badges and empty states.

## How did you test this change?

* Locally spun up the production Docker environment (`docker-compose.prod.yml`).
* Configured Razorpay Live and Test API keys in the dashboard.
* Triggered manual "Sync Payments" and verified that Razorpay Catalog, Customers, Subscriptions, and Payment History all successfully pulled into the local database and populated the respective dashboard tables.
* Confirmed that missing keys gracefully show the "Configure Provider" empty-state UI without breaking the sync loop for other environments.
* Tested the frontend to ensure that the Stripe and Razorpay empty states and UI badges are consistent in both light and dark modes.
#1347 
#1368 
Screenshot 2026-05-28 at 9 06 43 PM
Screenshot 2026-05-28 at 9 06 59 PM
Screenshot 2026-05-28 at 9 07 28 PM
Screenshot 2026-05-28 at 9 07 50 PM
Screenshot 2026-05-28 at 9 08 07 PM
Screenshot 2026-05-28 at 9 08 29 PM
Screenshot 2026-05-28 at 9 08 46 PM

---
## Summary by cubic
Adds complete Razorpay support alongside Stripe with provider-scoped data, verified webhooks, and full sync across environments. Implements Linear #1347 and #1368.

- **New Features**
  - Backend: Razorpay provider with pagination; `/api/payments/razorpay/*` routes; config, sync, and webhook services; webhooks return 200 and process async using `req.rawBody`.
  - Database: new `payments.razorpay_connections` and `payments.razorpay_webhook_events`; add `provider` to shared payment tables (default `'stripe'`); make `stripe_customer_id` nullable.
  - Frontend: Payments Settings adds Razorpay keys/webhooks and manual sync; `useRazorpay*` hooks; `ProviderBadge`; tables show provider and load when either provider is configured.
  - Dependencies & Env: add `razorpay` SDK and `RAZORPAY_*` keys in `.env.example`; Dockerfile dev stage moved so runner is default on Railway.

- **Bug Fixes**
  - Webhooks: validate HMAC signatures with raw bytes; non-blocking handler returns 200 immediately; restore dedup window; compute collision-resistant idempotency keys.
  - Sync: single-flight guard for concurrent Razorpay syncs; provider-scoped deletes; remove invalid empty-array guard; preserve partial failures for `all`.
  - Status/Config: sequential `getStatus` with try/catch; transactional upserts for keys/webhooks; key-type discriminator for Razorpay keys; fix dashboard filtering and provider badge case sensitivity; queries now select explicit `provider`; prefer latest sync timestamp.
  - Catalog: include missing `provider` in listCatalog query and schema to support badges.
  - Migrations: renumbered and hardened with CHECKs and safer deletes; keep `provider` DEFAULT `'stripe'` to avoid Stripe insert failures.

<sup>Written for commit 2389ea8ae0da7037b971061b003afc02a1a96de9. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/InsForge/InsForge/pull/1382?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

> [!NOTE]
> ### Add Razorpay as a payment provider alongside Stripe across backend and dashboard
> - Adds a full Razorpay integration: backend `RazorpayProvider`, `RazorpayConfigService`, `RazorpaySyncService`, and `RazorpayWebhookService` handle credential management, paginated data sync, and webhook deduplication.
> - Exposes new admin REST endpoints under `/api/payments/razorpay` for status, config CRUD, webhook setup, and sync, consumed by a new `razorpayService` client in the dashboard.
> - Adds database migrations (051–054) to create `razorpay_connections` and `razorpay_webhook_events` tables and add a `provider` discriminator column (default `'stripe'`) to all payment entity tables.
> - Dashboard pages (Catalog, Customers, Payment History, Subscriptions) now show a `ProviderBadge` per row and a settings dialog with a new Razorpay Keys tab and side-by-side Stripe/Razorpay sync and webhook sections.
> - Risk: existing payment records get `provider = 'stripe'` via migration defaults; shared schemas now require a `provider` field on products, prices, customers, subscriptions, and payment history, which is a breaking schema change for any consumers not yet updated.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 2389ea8.</sup>
> 
>

## Summary by CodeRabbit

* **New Features**
  * Added Razorpay support: per-environment keys, webhook setup/verification, idempotent webhook handling, and full sync (catalog, customers, subscriptions, payments) with dashboard controls.
  * Dashboard: Razorpay UI in Settings, key-missing state, provider badges/columns, and provider-aware sync/webhook actions.
* **Chores**
  * Made Stripe subscription customer ID nullable.
* **Documentation**
  * .env example updated with Razorpay credential placeholders.

### CODE REVIEW

File: backend/src/api/routes/payments/razorpay.routes.ts [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+  // Wait for sync to complete before acknowledging the webhook
+  try {
+    await syncService.syncAll(environment);
+  } catch (err) {
+    const message = err instanceof Error ? err.message : String(err);
+    console.error(`[Razorpay Webhook] Background sync failed for ${environment}: ${message}`);
+    throw err;
+  }
+
+  return true;
```
[the remark this answers is not in the document]
Comment (hemanthreddykoduru) [reply backend/src/api/routes/payments/razorpay.routes.ts]: Update: All CI checks resolved ✅
Addressed all review feedback and CI failures:

Backend fixes:

Fixed webhook HMAC signature verification to use raw request bytes instead of JSON.stringify()
Added timingSafeEqual buffer length guard to prevent crashes on malformed signatures
Fixed transaction atomicity in setRazorpayKeys using dedicated pool.connect() client
Implemented proper pagination (while loops with skip/count) in all RazorpayProvider list methods (listPlans, listItems, listCustomers)
Replaced Promise.all with sequential try-catch in getStatus() so Razorpay errors don't break Stripe status
Added pending reclaim deduplication window (5 min) for Razorpay webhook events (matching Stripe's pattern)
Scoped subscription cleanup queries with raw->>'object'/raw->>'entity' filters to prevent cross-provider row deletion
Fixed all TypeScript compilation errors (as unknown as casts for Razorpay SDK responses, missing RazorpayConnection import)
Dashboard fixes:

Removed unused React import in ProviderBadge.tsx
Fixed CatalogPage to show the most recent sync timestamp across both providers
Fixed provider badge detection in SubscriptionsPage using raw.entity instead of customer ID prefix heuristic
Fixed Razorpay sync to properly surface partial failures instead of silently swallowing them
Formatting:

Resolved all Prettier formatting errors across backend and dashboard packages
All lint, format, and typecheck checks now pass cleanly

**Comment by hemanthreddykoduru** [comment]
Hey @Fermionic-Lyu , all the CI checks are passing now and I've verified the integration works end-to-end locally. Let me know if you need any changes! #1382

**Comment by hemanthreddykoduru** [comment]
Hey @Fermionic-Lyu, the CI was failing because of a migration number clash with some newly merged PRs in main. I've renumbered the Razorpay migrations to 051, 052, and 053 and now all checks are passing perfectly. Ready for review when you have time!

File: backend/src/services/payments/razorpay-sync.service.ts [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+  // ─── Catalog: Plans → products table ─────────────────────────────────────────
+
+  private async upsertPlansAsProducts(
+    client: PoolClient,
+    environment: RazorpayEnvironment,
+    plans: RazorpayPlan[]
+  ): Promise<void> {
+    for (const plan of plans) {
+      await client.query(
+        `INSERT INTO payments.products (
```
Comment (Fermionic-Lyu) [inline backend/src/services/payments/razorpay-sync.service.ts:202]: This is the main blocker for me: Razorpay is being written into the same Stripe-oriented mirror tables and `stripe_*` columns. The existing Stripe flows still query and clean up mostly by `environment` only, so Stripe sync can remove/mutate Razorpay rows and existing Stripe list APIs can return Razorpay records. Please add a real provider discriminator, or split provider-owned data, and make every list/sync/cleanup path explicitly provider-scoped. Relying on ID prefixes or `raw` shape is too fragile.

File: backend/src/api/routes/payments/razorpay.routes.ts [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+      const payload = req.body as RazorpayWebhookPayload;
+      const eventStart = await webhookService.recordWebhookEventStart(environment, payload);
+
+      if (!eventStart.shouldProcess) {
+        res.status(200).json({ received: true, handled: false });
+        return;
+      }
+
+      // Route to specific handlers based on event type
+      const handled = await handleRazorpayWebhookEvent(environment, payload);
```
Comment (Fermionic-Lyu) [inline backend/src/api/routes/payments/razorpay.routes.ts]: This webhook path waits for a full `syncAll(environment)` before returning HTTP 200. A full sync can paginate plans, items, customers, subscriptions, and payments, then write several DB transactions, so this can exceed Razorpay webhook delivery timeouts and cause retries or disabled delivery. Also, `syncAll()` returns failed result objects in several cases instead of throwing, but this route still marks the webhook event as processed. Please acknowledge quickly after signature/idempotency checks, run sync in the background or a job queue, and mark failed sync results as failed.

File: backend/src/services/payments/razorpay-sync.service.ts [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+          );
+        }
+      }
+
+      // Clean up subscriptions that are no longer in Razorpay
+      const syncedIds = subscriptions.map((s) => s.id);
+      await client.query(
+        `DELETE FROM payments.subscription_items
+         WHERE environment = $1
+           AND (raw->>'entity' = 'subscription_item' OR raw->>'entity' = 'item')
```
Comment (Fermionic-Lyu) [inline backend/src/services/payments/razorpay-sync.service.ts]: The cleanup predicate does not match the rows inserted above. Razorpay subscription items store `raw = sub`, whose `entity` is `subscription`, not `subscription_item` or `item`, so stale Razorpay subscription items will not be deleted. Once provider scoping is added, this cleanup should delete by provider + environment + parent subscription IDs rather than this `raw.entity` check.

File: packages/dashboard/src/features/payments/services/razorpay.service.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        this.syncPayments({ environment: 'live' }),
+      ]);
+
+      const successes = results
+        .filter((r): r is PromiseFulfilledResult<any> => r.status === 'fulfilled')
+        .map((r) => r.value);
+
+      const failures = results.filter((r): r is PromiseRejectedResult => r.status === 'rejected');
+
+      if (failures.length > 0) {
```
Comment (Fermionic-Lyu) [inline packages/dashboard/src/features/payments/services/razorpay.service.ts]: The hook has toast logic for partial failures, but this throws when either environment fails, discarding any successful environment result. That means the partial-failure UX cannot work and successful syncs may not refresh correctly. Please return the fulfilled results and include failed environment state in the response/toast path instead of throwing on the first rejected environment.

**Review by Fermionic-Lyu — CHANGES_REQUESTED** [review]
Thanks a lot for implementing Razorpay. I really like this feature, and we do want to support Razorpay as soon as we can. Before merging, though, please make sure the two providers cannot conflict with each other. For the frontend UI, I will ask our designer to design the Razorpay-specific experience, and I can help with the dashboard side.

**Comment by hemanthreddykoduru** [comment]
@Fermionic-Lyu The PR is fully updated and ready for another look! I've addressed all of your blockers, and also went ahead and fixed the edge cases caught by the bots.

Here is a summary of the latest fixes:

1. Provider Discriminator (Main Blocker Fixed): I added migration 054 which introduces a strictly typed provider column across all shared payment mirror tables. The tables, list APIs, and sync/cleanup paths all now explicitly filter by provider = 'razorpay'. The migration safely maintains DEFAULT 'stripe' under the hood so that the legacy Stripe sync inserts don't break.

2. Webhook Timeouts & Status Tracking: The webhook handler now responds with a 200 OK immediately after signature verification to avoid Razorpay's 5-second timeout, pushing the syncAll workload to the background. Bot Fix: We now intentionally leave the webhook event as pending while the background sync runs. We only mark it as processed (or failed with the exact error) after the sync promise resolves, ensuring transient failures are actually tracked and retryable.

3. Webhook Idempotency Key Collisions: Bot Fix: Updated the idempotency key logic. If Razorpay fires two webhooks in the exact same second, our key now injects the specific entity ID (e.g., pay_xxxx from the payload) so they no longer collide and silently drop.

4. Concurrent Sync Exhaustion: Bot Fix: Added a single-flight mutex guard (syncPromises Map) in RazorpaySyncService. If a burst of webhooks hits for the same environment, it deduplicates the background jobs so we don't bombard the Razorpay API or exhaust the DB connection pool.

5. Partial Sync Errors Preserved: Instead of throwing on the first rejected environment, the frontend hook and backend service now return fulfilled/rejected results per environment. This preserves successful syncs while cleanly showing toast errors for the specific environments that failed.

All CI tests and linters are completely green. Let me know if you need anything else!

**Comment by hemanthreddykoduru** [comment]
Just pushed two small follow-up commits to fix an edge case where the Razorpay catalog badges weren't displaying on the frontend due to a missing provider column in the listCatalog SQL query. Everything is fully resolved and tested locally! Thanks again. @Fermionic-Lyu

**Comment by hemanthreddykoduru** [comment]
@Fermionic-Lyu I have implemented all the requested changes, added the missing provider column, and fully tested the Razorpay webhook orchestration in a live Railway production environment! Everything is working perfectly. I also included a small fix to the Dockerfile to make Railway deployments smoother. Let me know if you need anything else!

**Review by hemanthreddykoduru — COMMENTED** [review]
fixed

**Comment by hemanthreddykoduru** [comment]
All issues is fixed @Fermionic-Lyu please check it!!

**Comment by Fermionic-Lyu** [comment]
> All issues is fixed @Fermionic-Lyu please check it!!

Do you mind join our discord channel or contact me on discord? That might be easier for the following up work. I will help working on this PR and once finished, we need to test it so it works well for both Stripe & Razorpay, making sure no regressions on the existing behaviour. Since it looks like you have a Razorpay account, it would be great if you can help testing it. LMK if you can help on this, bunch of thanks!

**Comment by hemanthreddykoduru** [comment]
> > All issues is fixed @Fermionic-Lyu please check it!!
> 
> Do you mind join our discord channel or contact me on discord? That might be easier for the following up work. I will help working on this PR and once finished, we need to test it so it works well for both Stripe & Razorpay, making sure no regressions on the existing behaviour. Since it looks like you have a Razorpay account, it would be great if you can help testing it. LMK if you can help on this, bunch of thanks!

@Fermionic-Lyu can you please share your discord user name please

**Comment by hemanthreddykoduru** [comment]
> > All issues is fixed @Fermionic-Lyu please check it!!
> 
> Do you mind join our discord channel or contact me on discord? That might be easier for the following up work. I will help working on this PR and once finished, we need to test it so it works well for both Stripe & Razorpay, making sure no regressions on the existing behaviour. Since it looks like you have a Razorpay account, it would be great if you can help testing it. LMK if you can help on this, bunch of thanks!

"Absolutely! I would love to help get this over the finish line. I just joined the Discord. My username is [hemanthreddykoduru]. Let me know what you need me to test and I'll be happy to help make sure there are no regressions!"

**Review by Fermionic-Lyu — APPROVED** [review]
Merging to intermediate working branch
